x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/ElasticsearchInternalService.java

```diff
@@ -1004,4 +993,58 @@ static TaskType inferenceConfigToTaskType(InferenceConfig config) {
             return null;
         }
     }
+
+    // Iterates over the batch sending 1 request at a time to avoid
+    // filling the ml node inference queue.
+    class BatchIterator {
+        private final AtomicInteger index = new AtomicInteger();
+        private final ElasticsearchInternalModel esModel;
+        private final List<EmbeddingRequestChunker.BatchRequestAndListener> requestAndListeners;
+        private final InputType inputType;
+        private final TimeValue timeout;
+
+        BatchIterator(
+            ElasticsearchInternalModel esModel,
+            InputType inputType,
+            TimeValue timeout,
+            List<EmbeddingRequestChunker.BatchRequestAndListener> requestAndListeners
+        ) {
+            this.esModel = esModel;
+            this.requestAndListeners = requestAndListeners;
+            this.inputType = inputType;
+            this.timeout = timeout;
+        }
+
+        void doNextRequest() {
+            inferenceExecutor.execute(() -> inferOnBatch(requestAndListeners.get(index.get())));
+        }
+
+        private void inferOnBatch(EmbeddingRequestChunker.BatchRequestAndListener batch) {
```

Comment:
I don't think `inferOnBatch` is the right name. That doesn't capture the fact that it's subsequently doing the next batch as well, and the next till there all done.